### PR TITLE
refactor: simplify HTMLElement type assertions

### PR DESCRIPTION
**Pull request by maxpatiiuk**
- Improve the typing of "closestElementCrossShadowBoundary" to reduce need for type assertions.
- Remove needless type assertions.
  - You could also install this ESLint rule to detect these automatically: https://typescript-eslint.io/rules/no-unnecessary-type-assertion/
- Reduce usages of type assertions in favor of safer syntax

### CODE REVIEW

File: packages/calcite-components/src/components/action-bar/action-bar.tsx [hunk 1]
```diff
@@ -357,9 +357,9 @@ export class ActionBar
   }
 
   handleDefaultSlotChange = (event: Event): void => {
-    const groups = slotChangeGetAssignedElements(event).filter((el) =>
-      el.matches("calcite-action-group"),
-    ) as HTMLCalciteActionGroupElement[];
+    const groups = slotChangeGetAssignedElements(event).filter(
+      (el): el is HTMLCalciteActionGroupElement => el.matches("calcite-action-group"),
```
Comment (jcfranco) [inline packages/calcite-components/src/components/action-bar/action-bar.tsx:361]: Oooh, I like this. #TIL

File: packages/calcite-components/src/utils/dom.ts [hunk 2]
```diff
@@ -204,10 +204,19 @@ export function queryElementRoots<T extends Element = Element>(
...
+): HTMLElementTagNameMap[TagName] | null;
+export function closestElementCrossShadowBoundary<T extends Element = Element>(
+  element: Element,
+  selector: string,
+): T | null;
 export function closestElementCrossShadowBoundary<T extends Element = Element>(
   element: Element,
   selector: string,
 ): T | null {
+  document.createElement;
```
Comment (jcfranco) [inline packages/calcite-components/src/utils/dom.ts:219]: This seems unrelated. Is it intentional?

Comment (maxpatiiuk) [reply packages/calcite-components/src/utils/dom.ts]: Indeed - I used it for testing, but forgot to remove - thanks for catching

File: packages/calcite-components/src/utils/dom.ts [hunk 3]
```diff
@@ -200,20 +200,27 @@ export function queryElementRoots<T extends Element = Element>(
...
+export function closestElementCrossShadowBoundary<T extends Element = Element>(
+  element: Element,
+  selector: string,
+): T | null;
 export function closestElementCrossShadowBoundary<T extends Element = Element>(
   element: Element,
   selector: string,
 ): T | null {
-  // based on https://stackoverflow.com/q/54520554/194216
-  function closestFrom<T extends Element = Element>(el: Element): T | null {
```
Comment (maxpatiiuk) [inline packages/calcite-components/src/utils/dom.ts:212]: It appears that creating this function was needless.

We can get a small performance boost by not creating functions on the fly (as then we can benefit from v8 caching). More info:

https://marvinh.dev/blog/speeding-up-javascript-ecosystem/#:~:text=Inline%20functions%2C%20inline%20caches%20and%20recursion

Comment (jcfranco) [reply packages/calcite-components/src/utils/dom.ts:212]: Nice cleanup! The maintainability win is a great bonus too.

**Review by jcfranco — APPROVED** [review]
🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹
🧹👍🧹🧹🧹👍🧹👍👍👍🧹🧹👍👍👍🧹👍👍👍👍🧹👍🧹
🧹👍👍🧹🧹👍🧹🧹👍🧹🧹👍🧹🧹🧹🧹👍🧹🧹🧹🧹👍🧹
🧹👍🧹👍🧹👍🧹🧹👍🧹🧹👍🧹🧹🧹🧹👍👍👍🧹🧹👍🧹
🧹👍🧹🧹👍👍🧹🧹👍🧹🧹👍🧹🧹🧹🧹👍🧹🧹🧹🧹🧹🧹
🧹👍🧹🧹🧹👍🧹👍👍👍🧹🧹👍👍👍🧹👍👍👍👍🧹👍🧹
🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹🧹
